### PR TITLE
Add OpenSSL as supported licence type

### PR DESCRIPTION
**Pull request by cbracken**
Adds special handling for the OpenSSL licensed files under BoringSSL.
Specifically, some of the ARM assembly files in the latest Dart SDK are
dual-licensed with OpenSSL.

Those files (e.g., sha256-armv4.S) include a block that specifies the
code is dual-licenced under OpenSSL, but doesn't include a copyright
date of its own.

### CODE REVIEW

File: tools/licenses/lib/licenses.dart [hunk 1]
```diff
@@ -752,13 +756,16 @@ Iterable<_LicenseMatch> _tryReferenceByFilename(String body, LicenseFileReferenc
...
-  for (_PartialLicenseMatch match in _findLicenseBlocks(body, pattern, 1, 2)) {
+Iterable<_LicenseMatch> _tryReferenceByType(String body, RegExp pattern, LicenseSource parentDirectory, { String origin, bool needsCopyright: true }) sync* {
+  for (_PartialLicenseMatch match in _findLicenseBlocks(body, pattern, 1, 2, needsCopyright: needsCopyright)) {
     final LicenseType type = convertLicenseNameToType(match.group(3));
     final License template = parentDirectory.nearestLicenseOfType(type);
     if (template == null)
       throw 'failed to find accompanying $type license in $parentDirectory';
-    assert(_reformat(match.getCopyrights()) != '');
+    if (needsCopyright)
+      assert(_reformat(match.getCopyrights()) != '');
```
Comment (Hixie) [inline tools/licenses/lib/licenses.dart]: these should be throws, not asserts, so they throw in non-checked mode too

Comment (Hixie) [inline tools/licenses/lib/licenses.dart]: unless you really mean to just assert, in which case the whole four lines should all be in an assert block.

Comment (Hixie) [inline tools/licenses/lib/licenses.dart]: (yeah looks like this should just be an assert)

File: tools/licenses/lib/main.dart [hunk 2]
```diff
@@ -1993,13 +1993,40 @@ class RepositoryBoringSSLSourceDirectory extends RepositoryDirectory {
...
   RepositoryDirectory createSubdirectory(fs.Directory entry) {
     if (entry.name == 'third_party')
       return new RepositoryBoringSSLThirdPartyDirectory(this, entry);
     return super.createSubdirectory(entry);
   }
 }
 
+class RepositoryOpenSSLLicenseFile extends RepositorySingleLicenseFile {
+  RepositoryOpenSSLLicenseFile(RepositoryDirectory parent, fs.TextFile io)
+    : super(parent, io, new License.message(io.readString(), LicenseType.openssl, origin: io.fullName)) {
```
Comment (Hixie) [inline tools/licenses/lib/main.dart]: This is the place where we should trim off the first N lines (whatever number of lines the lawyer said to trim off)

File: travis/licenses.golden [hunk 3]
```diff
@@ -15254,7 +15254,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ====================================================================================================
 LIBRARY: boringssl
 ORIGIN: ../../../third_party/boringssl/src/LICENSE
-TYPE: LicenseType.unknown
+TYPE: LicenseType.openssl
 FILE: ../../../third_party/boringssl/src/crypto/bn/asm/x86_64-gcc.c
 FILE: ../../../third_party/boringssl/src/crypto/err/asn1.errordata
 FILE: ../../../third_party/boringssl/src/crypto/err/bio.errordata
```
Comment (Hixie) [inline travis/licenses.golden]: Travis additionally noticed that when we're looking for a license via this new path, two files end up here instead of a block that has no license text to reproduce. Can you verify that the other files in the block that these two files used to be in are correctly being given classified as not needing a license reproduced in the license screen? (It's quite possible, but better safe than sorry.)

**Comment by Hixie** [comment]
LGTM

File: travis/licenses.golden (hunk 3 above)
Comment (cbracken) [inline travis/licenses.golden]: Ah yep - that was me getting overly excited and pushing that second commit in that adds support for the new no-copyright block but, in my excitement, failing to update the golden file. Turns out the block in the second commit is in the existing version of that file. 

The upcoming Dart SDK roll adds a further OpenSSL block that I'll add handling for in the next PR.

File: tools/licenses/lib/licenses.dart (hunk 1 above)
Comment (cbracken) [inline tools/licenses/lib/licenses.dart]: Done.

File: tools/licenses/lib/main.dart (hunk 2 above)
Comment (cbracken) [inline tools/licenses/lib/main.dart]: Thanks! Will do this in a dedicated PR.

Comment (cbracken) [inline tools/licenses/lib/main.dart]: Done in #3347
